compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp

```diff
@@ -441,6 +461,92 @@ static void collectTiledAndFusedOps(Operation *rootOp,
   }
 }
 
+/// Consider the following case
+///
+/// ```mlir
+/// %0:2 = linalg.generic {
+///     indexing_maps = [....,
+///                      affine_map<(d0, d1, d2) -> (d0, d1),
+///                      affine_map<(d0, d1, d2) -> (d0, d1)>]}
+/// %1 = linalg.generic ins(%0#0, %0#1) {
+///     indexing_maps = [affine_map<(d0, d1) -> (d0, d1),
+///                      affine_map<(d0, d1) -> (d0, d1)]}
+/// ```
+///
+/// After tiling the first op we get
+///
+/// ```
+/// %0:2 = scf.forall ... {
+///   %1:2 = linalg.generic {
+///       indexing_maps = [....,
+///                        affine_map<(d0, d1, d2) -> (d0, d1),
+///                        affine_map<(d0, d1, d2) -> (d0, d1)>]}
+///   }
+/// }
+/// %2 = linalg.generic ins(%0#0, %0#1) {
+///     indexing_maps = [affine_map<(d0, d1) -> (d0, d1),
+///                      affine_map<(d0, d1) -> (d0, d1)]}
+/// ```
+///
+/// Due to a quirk of the fusion of consumers, fusing this consumer into the
+/// loop results in
+///
+/// ```
+/// %0:2 = scf.forall ... {
+///   %1:2 = linalg.generic {
+///       indexing_maps = [....,
+///                        affine_map<(d0, d1, d2) -> (d0, d1),
+///                        affine_map<(d0, d1, d2) -> (d0, d1)>]}
+///   %2 = tensor.extract_slice %0#1 [...]
+///   %3 = linalg.generic ins(%1#0, %2) {
+///       indexing_maps = [affine_map<(d0, d1) -> (d0, d1),
+///                        affine_map<(d0, d1) -> (d0, d1)]}
+///   }
+/// }
+/// ```
+///
+/// This is an SSA violation because of `%0#1` being used in the loop. This
+/// needs to be fixed upstream, but for cases where
+/// 1. The root operation produces results using an identity indexing map (when
+/// ignoring the iteration space dimensions corresponding to the reduction
+/// loops)
+/// 2. For all consumers of the results of the root operation, access the data
+/// using identity indexing map then for each consumer fusion step it is valid
+/// to replace all uses of slices of the outer loop that occur within the loop
+/// with the correponding tiled result value.
+/// This is a workaround till upstream transformation can fix this issue. The
+/// following method is testing if such a case exists to implement the
+/// work-around.
+static bool warForConsumerFusionSSAViolation(
+    Operation *rootOp,
+    const llvm::SmallDenseSet<Operation *> &tiledAndFusedOps) {
+  auto linalgRootOp = dyn_cast<linalg::LinalgOp>(rootOp);
+  if (!linalgRootOp) {
+    return false;
+  }
+  SmallVector<utils::IteratorType> iteratorTypes =
+      linalgRootOp.getIteratorTypesArray();
+  for (AffineMap map :
+       llvm::map_range(linalgRootOp.getIndexingMaps(), [](Attribute attr) {
+         return cast<AffineMapAttr>(attr).getValue();
+       })) {
```

Comment:
Is this to skip the SmallVector?
```suggestion
  for (AffineMap map : linalgRootOp.getIndexingMapsArray()) {
```

Reply:
True, but I am just trying to avoid materializing an array just to iterate over it. Something that of late annoys me quite a bit.